qiskit_experiments/data_processing/data_processor.py

```diff
@@ -125,79 +190,117 @@ def _call_internal(
-                    node.train(self._call_internal(data, call_up_to_node=index)[0])
+                    node.train(self._call_internal(data, call_up_to_node=index))
 
-    def _data_extraction(self, data: Union[Dict, List[Dict]]) -> List:
+    def _data_extraction(self, data: Union[Dict, List[Dict]]) -> np.ndarray:
         """Extracts the data on which to run the nodes.
 
         If the datum is a list of dicts then the data under self._input_key is extracted
-        from each dict and appended to a list which therefore contains all the data. If the
-        data processor has to_array set to True then the list will be converted to a numpy
-        array.
+        from each dict and appended to a list which therefore contains all the data.
 
         Args:
-            data: A list of such dicts where the data is contained under the key self._input_key.
+            data: A list of such dicts where the data is contained under the key
+                ``self._input_key``.
 
         Returns:
             The data formatted in such a way that it is ready to be processed by the nodes.
 
         Raises:
             DataProcessorError:
                 - If the input datum is not a list or a dict.
-                - If the data processor received a single datum but requires all the data to
-                  process it properly.
                 - If the input key of the data processor is not contained in the data.
+                - If the data processor receives multiple data with different measurement
+                  configuration, i.e. Jagged array.
         """
         if isinstance(data, dict):
             data = [data]
 
+        data_to_process = []
+        dims = None
+        for datum in data:
+            try:
+                outcome = datum[self._input_key]
+            except TypeError as error:
+                raise DataProcessorError(
+                    f"{self.__class__.__name__} only extracts data from "
+                    f"lists or dicts, received {type(data)}."
+                ) from error
+            except KeyError as error:
+                raise DataProcessorError(
+                    f"The input key {self._input_key} was not found in the input datum."
+                ) from error
+
+            if self._input_key != "counts":
+                outcome = np.asarray(outcome)
+                # Validate data shape
+                if dims is None:
+                    dims = outcome.shape
+                else:
+                    # This is because each data node creates full array of all result data.
+                    # Jagged array cannot be numerically operated with numpy array.
+                    if outcome.shape != dims:
+                        raise DataProcessorError(
+                            "Input data is likely a mixture of job results with different "
+                            "measurement setup. Data processor doesn't support jagged array."
+                        )
```

Comment:
It might be best to leave this conversion to the format method of the nodes? I don't think that the processor should be restricted to arrays. The nodes can impose the restriction but the processor should be agnostic to this. I wonder if the first node of the chain can tell the processor what format it would like?

Reply:
Previously this check doesn't exist. If we want this validation, we need to copy this code to most of nodes using axis or shape. Please check this example
```python
>>> arr = np.array([[1, 2], [1,2,3], [1,2,3,4]])
<ipython-input-78-4a6749cf8160>:1: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray.
  arr = np.array([[1, 2], [1,2,3], [1,2,3,4]])

>>> arr.shape
(3,)
```
As you can find, this will break most of validation logic. Or, we should use #474 (generator)